controller/model/asynchronous_eventer.go

```diff
@@ -0,0 +1,280 @@
+type EventListenerFunc func(name string, data []byte)
+
+// AsynchronousEventerBbolt implements AsynchronousEventer with a bbolt back storage mechanism.
+// Work is performed on a configurable basis via the Interval property in FIFO order.
+//
+// Events are stored in the following format:
+//		id   - CUID   - a monotonic reference id
+//      name - string - an event name, used for log output
+//      data - []byte - a string array of arguments
+type AsynchronousEventerBbolt struct {
+	handlerMap  cmap.ConcurrentMap //eventName -> handler func(name
+	BucketName  string
+	Interval    time.Duration
+	closeNotify <-chan struct{}
+	stopNotify  chan struct{}
+
+	env Env
+
+	running    concurrenz.AtomicBoolean
+	processing concurrenz.AtomicBoolean
+	batchSize  int
+}
+
+var _ AsynchronousEventer = &AsynchronousEventerBbolt{}
+
+// NewAsynchronousEventerBbolt creates a new bbolt backed asynchronous eventer that will check for new events at the given interval
+// or when triggered. On each interval/trigger, the number of events processed is determined by batchSize.
+func NewAsynchronousEventerBbolt(env Env, interval time.Duration, batchSize int) *AsynchronousEventerBbolt {
+	return &AsynchronousEventerBbolt{
+		Interval:  interval,
+		env:       env,
+		batchSize: batchSize,
+	}
+}
+
+func (a *AsynchronousEventerBbolt) AddEvent(eventType string, data []byte) {
+	err := a.env.GetDbProvider().GetDb().Update(func(tx *bbolt.Tx) error {
+		ctx := boltz.NewMutateContext(tx)
+		event := &persistence.AsynchronousEvent{
+			BaseExtEntity: boltz.BaseExtEntity{
+				Id:        cuid.New(),
+				CreatedAt: time.Now(),
+				UpdatedAt: time.Now(),
+			},
+			Type: eventType,
+			Data: data,
+		}
+
+		return a.env.GetStores().AsynchronousEvent.Create(ctx, event)
+	})
+
+	if err != nil {
+		pfxlog.Logger().WithError(err).Error("error adding event for AsynchronousEventerBbolt")
+	}
+}
+
+func (a *AsynchronousEventerBbolt) AddListener(eventType string, listener EventListenerFunc) {
+	a.handlerMap.Upsert(eventType, nil, func(exist bool, valueInMap interface{}, newValue interface{}) interface{} {
+		var handlers []EventListenerFunc
+		if exist {
+			handlers = valueInMap.([]EventListenerFunc)
+		}
+
+		handlers = append(handlers, listener)
+		return handlers
+	})
+}
+
+func (a *AsynchronousEventerBbolt) Start(closeNotify <-chan struct{}) error {
+	if !a.running.CompareAndSwap(false, true) {
+		return errors.New("already started")
+	}
+	a.stopNotify = make(chan struct{}, 0)
+	a.closeNotify = closeNotify
+
+	go a.run()
+
+	return nil
+}
+
+func (a *AsynchronousEventerBbolt) run() {
+
+	for {
+		select {
+		case <-a.closeNotify:
+			a.running.Set(false)
+			return
+		case <-a.stopNotify:
+			a.running.Set(false)
+			return
+		case <-time.After(a.Interval):
+			a.process()
+		}
+	}
+}
+
+func (a *AsynchronousEventerBbolt) Stop() error {
+	if a.running.CompareAndSwap(true, false) {
+		close(a.stopNotify)
+		return nil
+	}
+
+	return errors.New("not started")
+}
+
```

Comment:
An alternate approach would be to have a notification channel with space for 1 struct{}, which the run could listen to. Then the Trigger could try to push something onto the channel. If it's already full, then the process is already running. 
The main advantage to this approach is that you can trigger and not worry about the calling thread having to do processing. Thoughts?

Reply:
What happens if 3 processes all attempt to call `Trigger()` at once between interval runs? The first would start the process the next 2 would block on trying to send to the channel. I don't know if blocking on `Trigger()` is going to be a good idea.

Reply:
You do something like
```
select {
   case trigger <- struct{}:
   default:
}
```
If you can't submit to the trigger channel then someone has already done so and you can continue on your merry way.

Reply:
I'm ambivalent towards this - I think it will be fine either way. So I will implement the channel.

It does have the side effect of allowing 1 `Trigger` to be requested - which depending on when it is called may immediately pick up new events or cause a query for 0 events.